docs: update readme, docs layout and add user, dev docs [skip ci] (#87)
red-hat-storage/lvm-operator: docs/dev-guide/custom-resource-definitions.md

This page outlines `spec` and `status` fields of all the CRs managed by LVMO

## 1. LVMCluster

- LVMCluster CR represents the volume groups that should be created and managed
  across nodes with custom node selector, toleration and device selectors

> - Should be created and edited by user in operator installed namespace and only
>   single CR is supported, creation of multiple CRs is erroneous

Reviewer: ```suggestion
- Should be created and edited by user in operator installed namespace.
- only a single instance is supported at present.
```

Author: removing this file, same for rest of the comments in this file.

| Field    | Type             | Description                                                      |
|----------|------------------|------------------------------------------------------------------|
| `spec`   | LVMClusterSpec   | Desired behaviour of LVM Cluster with Volume Groups across Nodes |
| `status` | LVMClusterStatus | Observed status of LVM Cluster with Volume Groups across Nodes   |

### LVMClusterSpec

| Field           | Type                | Description                                                                        |
|-----------------|---------------------|------------------------------------------------------------------------------------|

> | `toleraions`    | []corev1.Toleration | List of tolerations that Topolvm Controller, Node Plugin and VG Manager can run on |

Reviewer: Typo: tolerations

> | `toleraions`    | []corev1.Toleration | List of tolerations that Topolvm Controller, Node Plugin and VG Manager can run on |

Reviewer: ```suggestion
| `tolerations`    | []corev1.Toleration | List of tolerations for the nodes that Topolvm Controller, Node Plugin and VG Manager can run on |
```

| `deviceClasses` | []DeviceClass       | List of deviceclasses                                                              |

#### DeviceClass

| Field            | Type                  | Description                           |
|------------------|-----------------------|---------------------------------------|
| `name`           | string                | Name of the Volume Group              |

> | `deviceSelector` | \*DeviceSelector      | Selector to match devices on the node |

Reviewer: ```suggestion
| `deviceSelector` | \*DeviceSelector      | Selector to match devices to be used |
```

| `nodeSelector`   | \*corev1.NodeSelector | Selector to match nodes               |

#### DeviceSelector

- Not Implemented, empty struct

### LVMClusterStatus

| Field                 | Type                | Description                           |
|-----------------------|---------------------|---------------------------------------|
| `ready`               | bool                | State of the LVMCluster               |
| `deviceClassStatuses` | []DeviceClassStatus | List of observed deviceClass statuses |

#### DeviceClassStatus

| Field        | Type         | Description                                          |
|--------------|--------------|------------------------------------------------------|
| `name`       | string       | Name of the device class                             |

> | `nodeStatus` | []NodeStatus | Status of the node that belongs to this device class |

Reviewer: ```suggestion
| `nodeStatus` | []NodeStatus | Status of the VG on the nodes |
```

#### NodeStatus

| Field     | Type         | Description                                                           |
|-----------|--------------|-----------------------------------------------------------------------|
| `Node`    | string       | Name of the node                                                      |
| `status`  | VGStatusType | One of `Ready`, `Failed`, `Degraded`                                  |
| `devices` | []string     | List of Physical Volumes (devices) that are part of this device class |

> ### Lifecycle

Reviewer: Move this out of this file and to the deployment and user guides.

1. Only a single volume group containing all available disks across schedulable
   nodes is supported and implementation respecting tolerations, device andi
   node selector fields is coming soon

> 2. All fields in `status` is updated based on the status of volume groups

Reviewer: ```suggestion
2. All fields in `status` are updated based on the status of volume groups
```

   creation across nodes

> 3. `LVMCluster` is created with a [finalizer][finalizer] and on deletion of CR,

Reviewer: Not required in this file. It is an implementation detail, not a CRD detail.

   [entities][reconciler] created by LVMO will be deleted

## 2. LVMVolumeGroup

| Field    | Type                 | Description                                 |
|----------|----------------------|---------------------------------------------|

> | `spec`   | LVMVolumeGroupSpec   | Desired Volume Groups on respective nodes   |

Reviewer: ```suggestion
| `spec`   | LVMVolumeGroupSpec   | Specification of a single Volume Groups   |
```

> | `status` | LVMVolumeGroupStatus | Status of Volume Groups on respective nodes |

Reviewer: ```suggestion
| `status` | LVMVolumeGroupStatus | Status of the Volume Group |
```

### LVMClusterSpec

| Field            | Type                  | Description                           |
|------------------|-----------------------|---------------------------------------|
| `deviceSelector` | \*DeviceSelector      | Selector to match devices on the node |

> | `nodeSelector`   | \*corev1.NodeSelector | Selector to match nodes               |

Reviewer: ```suggestion
| `nodeSelector`   | \*corev1.NodeSelector | Selector to determine nodes on which the volume group should be created            |
```

### LVMVolumeGroupStatus

- Not Implemented, empty struct

### Lifecycle

> 1. Created and managed by LVMO, this contains subset of fields from LVMCluster
>    CR. Volume Groups with same name can exist on different nodes and to
>    reconcile based on this CR reduces the burden for multiple controllers
> 

Reviewer: ```suggestion
1. The LVMO creates an LVMVolumeGroup CR for each deviceClass in the LVMCluster CR. The LVMVolumeGroups are reconciled by the vgmanager controllers.
```
